pkg/credentialprovider/plugin/config.go

```diff
@@ -43,6 +45,15 @@ func readCredentialProviderConfigFile(configPath string) (*kubeletconfig.Credent
 		return nil, fmt.Errorf("error decoding config %s: %w", configPath, err)
 	}
 
+	// Append current system environment variables, to the ones configured in the
+	// credential provider file. Failing to do so may result in unsuccessful execution
+	// of the provider binary, see https://github.com/kubernetes/kubernetes/issues/102750
+	// Also, this behaviour is inline with Credential Provider Config spec
+	systemEnvVars := os.Environ()
```

Comment:
Should we instead call `os.Environ()` and append it to `cmd.Env` at runtime instead of at start up? 

https://github.com/kubernetes/kubernetes/blob/9c7434bc4f30761cd679b96fd818aa68a9727d5d/pkg/credentialprovider/plugin/plugin.go#L348-L351

Reply:
I'm guessing we're calling it at startup so it only needs to be loaded once?

Reply:
> @adisky Sure makes sense and I was under a dilemma of doing it in `ExecPlugin` v/s `readCredentialProviderConfigFile` and I was hoping to get some feedback on it. 
> 
> 
> 
> My rationale was providing a `CredentialProviderConfig` is the job of `readCredentialProviderConfigFile` and all other functions are consumers of it.
> 
> 
> 
> If I am appending env vars in the `ExecPlugin` function then it's implied that `ExecPlugin` knows that env vars could be stale and this breaks the abstraction between `ExecPlugin` and `readCredentialProviderConfigFile`
> 
> 
> 
> At the moment, `ExecPlugin` does is just take the env vars present in the `CredentialProviderConfig` and execute it and I didn't want to break this contract.
> 
> 
> 
> I'll wait for the feedback from the other reviewers and if general consensus is to move it to `ExecPlugin` to inject latest env vars without having to perform a restart of the kubelet then I'll do it.

@andrewsykim Above is my reason for putting it in readCredentialProviderConfigFile let me know your thoughts on this?

Reply:
> At the moment, ExecPlugin does is just take the env vars present in the CredentialProviderConfig and execute it and I didn't want to break this contract.

Automatically injecting env vars into at start up has the equally suprising effect I think. From the consumer's perpective it is the same because it wasn't initially defined in `CredentialProviderConfig`. 

I'm personally on the fence about whether inheriting all of kubelet's env vars is the right approach, as opposed to requiring `CredentialProviderConfig` to explicitly set all the required values. Will think about this some more

Reply:
@andrewsykim I believe the behaviour of unioning system's env vars with config provided env var is part of the publicly documented [spec](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go#L556) and there has to be some thought process in place behind this decision.

To make things more transparent I can log the env vars which are being passed to the credential binary and maybe I can even differentiate between system vars and config vars, this should help in troubleshooting if required.

Moreover, in my implementation config provided env vars always take priority over system provided ones giving a level of control to the end user

If it makes sense then you and I can have an offline discussion about it :)

Reply:
> To make things more transparent I can log the env vars which are being passed to the credential binary and maybe I can even differentiate between system vars and config vars, this should help in troubleshooting if required.

This can be dangerous too since we can end up logging credentials.

> @andrewsykim I believe the behaviour of unioning system's env vars with config provided env var is part of the publicly documented spec and there has to be some thought process in place behind this decision.

This is fair, but the API is also alpha right now so we _can_ still change it.

Reply:
@andrewsykim Makes sense.

Do you want some time to think over the approach of unioning the vars?

One solution which I can see to handle issues like #102750 is to maybe only pass HOME and PATH env vars and document this behaviour.

Reply:
Hello @andrewsykim did you get a chance to look at the above?

Reply:
@andrewsykim Did you get a chance to look at the above comments?

Reply:
Since we're likely not going to be promoting this feature to Beta this release I think we should defer this for v1.23 and revisit for Beta. Thoughts @adisky ?

Reply:
unioning/overlaying the default os envvars is what I expected (and is what client-go exec auth plugin does)

Reply:
see

https://github.com/kubernetes/kubernetes/blob/e0a22acaa0c62f3e6f9dd37ab2a4e7d960528edc/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L201

https://github.com/kubernetes/kubernetes/blob/e0a22acaa0c62f3e6f9dd37ab2a4e7d960528edc/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L419

Reply:
^ thanks for the reference Jordan! Probably makes sense to follow same patterns as the client-go exec auth plugin in this case.

Reply:
@andrewsykim Now that we have decided to take it forward as a bug fix v1.22, are we good with the code changes or you would like me to refine it further?